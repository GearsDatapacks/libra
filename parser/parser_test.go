package parser_test

import (
	"testing"

	"github.com/gearsdatapacks/libra/lexer"
	"github.com/gearsdatapacks/libra/lexer/token"
	"github.com/gearsdatapacks/libra/parser"
	"github.com/gearsdatapacks/libra/parser/ast"
	utils "github.com/gearsdatapacks/libra/test_utils"
)

func getProgram(t *testing.T, input string) *ast.Program {
	l := lexer.New(input, "test.lb")
	tokens := l.Tokenise()

	p := parser.New(tokens, l.Diagnostics)
	program := p.Parse()
	utils.AssertEq(t, len(p.Diagnostics.Diagnostics), 0)

	return program
}

func getExpr[T ast.Expression](t *testing.T, program *ast.Program) T {
	utils.AssertEq(t, len(program.Statements), 1)
	stmt, ok := program.Statements[0].(*ast.ExpressionStatement)
	utils.Assert(t, ok)

	expr, ok := stmt.Expression.(T)
	utils.Assert(t, ok)
  return expr
}

func TestIdentifierExpression(t *testing.T) {
	input := "hello_World123"
  program := getProgram(t, input)

  ident := getExpr[*ast.Identifier](t, program)

	utils.AssertEq(t, ident.Name, input)
	utils.AssertEq(t, ident.Token,
		token.New(token.IDENTIFIER, input,
			token.NewSpan(0, 0, len(input)),
		),
	)
}

func TestIntegerExpression(t *testing.T) {
	input := "156_098"
	val := 156098

  program := getProgram(t, input)

  integer := getExpr[*ast.IntegerLiteral](t, program)

	utils.AssertEq(t, integer.Value, int64(val))
	utils.AssertEq(t, integer.Token,
		token.New(token.INTEGER, "156098",
			token.NewSpan(0, 0, len(input)),
		),
	)
}

func TestFloatExpression(t *testing.T) {
	input := "3.1415"
	val := 3.1415

  program := getProgram(t, input)

  float := getExpr[*ast.FloatLiteral](t, program)

	utils.AssertEq(t, float.Value, val)
	utils.AssertEq(t, float.Token,
		token.New(token.FLOAT, input,
			token.NewSpan(0, 0, len(input)),
		),
	)
}

func TestBooleanExpression(t *testing.T) {
	input := "true"

  program := getProgram(t, input)

  boolean := getExpr[*ast.BooleanLiteral](t, program)

	utils.AssertEq(t, boolean.Value, true)
	utils.AssertEq(t, boolean.Token,
		token.New(token.IDENTIFIER, input,
			token.NewSpan(0, 0, len(input)),
		),
	)
}

func TestErrorExpression(t *testing.T) {
  input := ")"

  l := lexer.New(input, "test.lb")
  tokens := l.Tokenise()
  p := parser.New(tokens, l.Diagnostics)
  program := p.Parse()

  utils.AssertEq(t, len(p.Diagnostics.Diagnostics), 1)
  diag := p.Diagnostics.Diagnostics[0]
  utils.AssertEq(t, diag.Message, "Expected expression, got `)`")
  utils.AssertEq(t, diag.Span, token.NewSpan(0, 0, 1))

  getExpr[*ast.ErrorExpression](t, program)
}

func TestMissingNewlineError(t *testing.T) {
  input := "1 2"

  l := lexer.New(input, "test.lb")
  tokens := l.Tokenise()
  p := parser.New(tokens, l.Diagnostics)
  p.Parse()

  utils.AssertEq(t, len(p.Diagnostics.Diagnostics), 1)
  diag := p.Diagnostics.Diagnostics[0]
  utils.AssertEq(t, diag.Message, "Expected newline after statement, got integer")
  utils.AssertEq(t, diag.Span, token.NewSpan(0, 2, 3))
}

