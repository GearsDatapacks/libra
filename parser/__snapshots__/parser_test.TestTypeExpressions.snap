
[`void` - 1]
IDENT void (0:4)
---

[`f32` - 1]
IDENT f32 (0:3)
---

[`Type` - 1]
IDENT Type (0:4)
---

[`i64[10]` - 1]
INDEX_EXPR (3:4)
├─IDENT i64 (0:3)
└─INT_LIT 10 (4:6)
---

[`bool[2]` - 1]
INDEX_EXPR (4:5)
├─IDENT bool (0:4)
└─INT_LIT 2 (5:6)
---

[`string[]` - 1]
INDEX_EXPR (6:7)
└─IDENT string (0:6)
---

[`Foo[]` - 1]
INDEX_EXPR (3:4)
└─IDENT Foo (0:3)
---

[`{string: i32}` - 1]
MAP_EXPR (0:1)
└─KEY_VALUE
  ├─IDENT string (1:7)
  └─IDENT i32 (9:12)
---

[`{bool: bool}` - 1]
MAP_EXPR (0:1)
└─KEY_VALUE
  ├─IDENT bool (1:5)
  └─IDENT bool (7:11)
---

[`(string, string, bool)` - 1]
TUPLE_EXPR (0:1)
├─IDENT string (1:7)
├─IDENT string (9:15)
└─IDENT bool (17:21)
---

[`(f64[], i8[])` - 1]
TUPLE_EXPR (0:1)
├─INDEX_EXPR (4:5)
│ └─IDENT f64 (1:4)
└─INDEX_EXPR (10:11)
  └─IDENT i8 (8:10)
---

[`type Func = fn(SomeType, OtherType): bool` - 1]
TYPE_DECL Func (0:4)
└─FUNC_TYPE (12:14)
  ├─PARAM
  │ └─TYPE_OR_IDENT SomeType
  ├─PARAM
  │ └─TYPE_OR_IDENT OtherType
  └─IDENT bool (37:41)
---

[`type Func = fn(): string` - 1]
TYPE_DECL Func (0:4)
└─FUNC_TYPE (12:14)
  └─IDENT string (18:24)
---

[`i8 | i16 | i32 | i64` - 1]
BIN_EXPR | (0:20)
├─BIN_EXPR | (0:14)
│ ├─BIN_EXPR | (0:8)
│ │ ├─IDENT i8 (0:2)
│ │ └─IDENT i16 (5:8)
│ └─IDENT i32 (11:14)
└─IDENT i64 (17:20)
---

[`string | bool` - 1]
BIN_EXPR | (0:13)
├─IDENT string (0:6)
└─IDENT bool (9:13)
---

[`*string` - 1]
PTR_TYPE (0:1)
└─IDENT string (1:7)
---

[`*mut i32` - 1]
PTR_TYPE mut (0:1)
└─IDENT i32 (5:8)
---

[`?f32` - 1]
OPTION_TYPE (0:1)
└─IDENT f32 (1:4)
---

[`?(string[])` - 1]
OPTION_TYPE (0:1)
└─PAREN_EXPR (1:2)
  └─INDEX_EXPR (8:9)
    └─IDENT string (2:8)
---

[`!u8` - 1]
PREFIX_EXPR `!` (0:1)
└─IDENT u8 (1:3)
---

[`!{u32: string}` - 1]
PREFIX_EXPR `!` (0:1)
└─MAP_EXPR (1:2)
  └─KEY_VALUE
    ├─IDENT u32 (2:5)
    └─IDENT string (7:13)
---

[`?({string: Value}[10])` - 1]
OPTION_TYPE (0:1)
└─PAREN_EXPR (1:2)
  └─INDEX_EXPR (17:18)
    ├─MAP_EXPR (2:3)
    │ └─KEY_VALUE
    │   ├─IDENT string (3:9)
    │   └─IDENT Value (11:16)
    └─INT_LIT 10 (18:20)
---

[`!Foo[]` - 1]
PREFIX_EXPR `!` (0:1)
└─INDEX_EXPR (4:5)
  └─IDENT Foo (1:4)
---

[`(i32 | f32)[] | i32 | f32` - 1]
BIN_EXPR | (11:25)
├─BIN_EXPR | (11:19)
│ ├─INDEX_EXPR (11:12)
│ │ └─PAREN_EXPR (0:1)
│ │   └─BIN_EXPR | (1:10)
│ │     ├─IDENT i32 (1:4)
│ │     └─IDENT f32 (7:10)
│ └─IDENT i32 (16:19)
└─IDENT f32 (22:25)
---
